### ISSUE TITLE
Pyrelational - logging to Wandb

### ISSUE THREAD

**Issue by jlotthammer** [RelationRx/pyrelational#196]
Hi, I mentioned this over in https://github.com/RelationRx/pyrelational/issues/38 but wanted to bump to an open issue. This is a very interesting and well put together project! I'm looking for documentation on how to log to Wandb with pyrelational. I found this post mentioning supporting that, but I couldn't find how to perform this in the current interface. Any guidance?

Thanks so much!

_Originally posted by @jlotthammer in https://github.com/RelationRx/pyrelational/issues/38#issuecomment-2415206571_

**Comment by thomasgaudelet** [comment]
Hi Jeffrey! Thanks a lot for the kind words! Could you expand on what you are interested in logging specifically? 

Briefly, at the moment, when using a Lightning-based model manager, one can simply pass a Logger to the Trainer to report the training metrics during each active learning step. However, pyrelational doesn't currently support a way to log the AL step metrics to wandb or other services. We do have on the roadmap an action to add support for external loggers in the Pipeline but it hasn't been implemented yet.

Hope this helps!

**Comment by thomasgaudelet** [comment]
Hi Jeffrey, thinking on it, a simple way to achieve this for the second use-case above would be to 

1. subclass the Pipeline (e.g. PipelineWithLogger) such that it takes a logger as input and stores it as an attribute 
2. overwrite the `compute_performance` method such that it calls `super().compute_performance` and then logs the desired metrics to wandb using the logger attribute.

Cheers!

**Comment by jlotthammer** [comment]
Hi, Thomas. 

Thanks for your fast reply! Yeah, my expectation was that I'd just be able to pass a typical pytorch logger through the trainer config, but although a run is initialized nothing seems to actually be logged in WandB. The logs seemingly populate if I go to a particular run, but nothing is graphically shown in any run nor can I add any plots.

I'm working on trying to identify _**why**_ that is happening in my scenario because if I understand correctly it seems like this should 'just work'. As a control, when I run without pyrelational - just the lightning model - it's logging just fine.  I'll update in this thread if I figure anything out.  If you have any thoughts please feel free to let me know. My testing code is fairly mundane here:


    wandb_logger = WandbLogger(project="active_learning_resnet1d")
    # Load your dataset
    df = pd.read_csv("fake_data.csv")
    sequences = df["Sequence"].tolist()
    targets = df["Target"].values

    # Create the dataset
    dataset = SequenceDataset(sequences, targets)

    # Split the dataset
    total_size = len(dataset)
    train_size = int(0.7 * total_size)
    val_size = int(0.15 * total_size)
    test_size = total_size - train_size - val_size

    train_ds, val_ds, test_ds = random_split(dataset, [train_size, val_size, test_size])
    train_indices = train_ds.indices
    val_indices = val_ds.indices
    test_indices = test_ds.indices

    # Model manager
    model_manager = LightningMCDropoutModelManager(
        model_class=ResNet1D,
        model_config={
            "block": ResidualBlock1D,
            "input_channels": 20,
            "layers": [2, 2, 2, 2],
            "learning_rate": 1e-3,
        },
        trainer_config={
            "accelerator": "gpu",
            "epochs": 50,
            "precision": "16-mixed",
            
             ########## Here I pass the logger to the trainer ##########
            "logger": wandb_logger,
            
            "monitor_metric_name": "val_loss",
            "use_early_stopping": True,
        },
    )`
    
    
And my model has training/val steps that look like this
    
```
def training_step(self, batch, batch_idx):
        sequences, targets = batch
        predictions = self.forward(sequences)
        loss = self.criterion(predictions, targets)
        r2 = self.r2_score(predictions, targets)
        self.log("train_loss", loss, prog_bar=True)
        self.log("train_r2", r2, prog_bar=True)
        return loss
        
def validation_step(self, batch, batch_idx):
        sequences, targets = batch
        predictions = self.forward(sequences)
        loss = self.criterion(predictions, targets)
        r2 = self.r2_score(predictions, targets)
        self.log("val_loss", loss, prog_bar=True)
        self.log("val_r2", r2, prog_bar=True)        
```

**Comment by jlotthammer** [comment]
Hi, Thomas.
> Briefly, at the moment, when using a Lightning-based model manager, one can simply pass a Logger to the Trainer to report the training metrics during each active learning step. 

As best I can tell, a logger wasn't actually being passed to the trainer. I've forked the repository and added a check in the default configs, so that it appropriately passes a logger [falls back to default CSVLogger located at `checkpoint_dir/lightning_logs` if none is provided. Matching lightning, if False is provided there are no logs are written] and it now is logging albeit there's some oddities to figure out - e.g., how to split runs so WandB doesn't overwrite the data for each run as the process repeats for the next cycle. I see what you're saying about subclassing the pipeline for the raw performance metrics. I may do this for my workload, thanks!

Unrelated, but I also made a minor tweak to enable persistent workers in the data loaders in model manager class. This gave me a pretty hefty performance boost on my synthetic data task. Happy to PR if these changes are useful.

**Comment by thomasgaudelet** [comment]
Hi Jeffrey, thanks for raising this! And good catch, we need to think what's the best solution for logging at both levels. I expect having the same logger for all steps is what create the issue where data is overwritten. We may need to come up with our own logger wrapper.

Happy to review any PR adding a feature that you found useful!

**Comment by jlotthammer** [comment]
Cool, it's relatively minor changes. I'll follow up when I have a minute.

Yeah, 100% that's the issue with logging all steps. For a custom logger wrapper are you envisioning [overriding the wandb logger](https://github.com/Lightning-AI/pytorch-lightning/blob/06a8d5bf33faf0a4f9a24207ae77b439354350af/src/lightning/pytorch/loggers/wandb.py#L50) from lightning somehow or are you thinking of something more general here?

**Comment by thomasgaudelet** [comment]
Awesome! Looking forward!

It's fairly rough thoughts at the moment, but it seems to me that we need some form of logger orchestrator that either
- ensures the names of metrics logged are edited at each iteration such that we log new metrics instead of overriding existing ones.
- instantiates/culls loggers such that we get a fresh one at each iteration.

It'd need to be a tad more generic than wandb as we should try to support generic loggers. I am leaning towards option 1, but would be great to get your thoughts on this as well.

**Comment by jlotthammer** [comment]
Yeah, I'm not too familiar with logging options outside of weights and biases to be honest. I think one thing to be cognizant of though is if you're changing the names of the logged metrics each time I think you'd also need to manage / change what you're using to save the model with as well - e.g., the lightning model managers use `monitor_metric_name` I believe.

Please correct if I'm wrong because the interactions between each of the classes is a bit complicated, but it looks to me like the way the model manager and pipeline work right now is by resetting the current model back to `None` after iterating through all the currently available labeled data. This then causes causes reinstantiations of the trainer/model if it is `None` If that's true, maybe option 2 isn't such a bad idea? Thats my take from reading this at least:

https://github.com/RelationRx/pyrelational/blob/060c0c3501e4fe77cfaee3a81d69d907f96f2cef/pyrelational/model_managers/abstract_model_manager.py#L47 

https://github.com/RelationRx/pyrelational/blob/060c0c3501e4fe77cfaee3a81d69d907f96f2cef/pyrelational/model_managers/lightning_model_manager.py#L91

and the pipeline:
https://github.com/RelationRx/pyrelational/blob/060c0c3501e4fe77cfaee3a81d69d907f96f2cef/pyrelational/pipeline/pipeline.py#L157

Note that the logger needs to be passed to the trainer anyway [at least in the case of using pytorch lightning for handling logging], so if the trainers being reinstantiated then I think you'd have to pass a logger there. This is the change I've made in my PR in #198.

I'm not sure how other loggers handle things but at least in wandb you could create a new name for the run and have the same metric names for everything. e.g., I'm imagining something like

```
    def setup_logger(self, iteration: int) -> WandbLogger:
        """Initialize WandBLogger for logging."""
        return WandbLogger(
            project="my_favorite_active_learning_run", 
            name=f"iteration{iteration}",
        )
```
Where the logger then gets passed to the trainer during the reinstantiation events already in place. So each iteration gets its own name for a run under the same project. This name feature may be WandB specific though, I'm not sure how other MLops platforms handle logging. Lastly, you'd probably want the final iteration of logging then to show not just the training metrics of that training, but also the total active learning metrics as a function of the labeled data.
